### ISSUE TITLE
WebSocket connection to 'ws://hostname.local:81/' failed: Could not decode a text frame as UTF-8.

### ISSUE THREAD

**Issue by VaAndCob** [Links2004/arduinoWebSockets#840]
I think there is UTF-8 encode inside the web socket.broadcastTXT() .  usually work fast and reliable but sending text frame, sometimes triggers "Could not decode a text frame as UTF-8" on the client and the socket connection is closed!
I dig down and found that some text messages send characters that are Asian or other languages,   and trigger the error.
Is there a way to sanitize or prevent wrong "UTF-8 encoding", so the library will automatically encode UTF-8 before sending it to the client?

**Comment by Links2004** [comment]
do you have UTF-8 in you string / char array that you send via `broadcastTXT`?
`broadcastTXT` does not alter you text / data by adding random strings.
are you sure you data is 0x00 terminated?


the length is detected via `strlen` if you use use `broadcastTXT` with out the length parameter.
https://github.com/Links2004/arduinoWebSockets/blob/751cf87b6cd684c9d339f0314a18b0ee866d449c/src/WebSocketsServer.cpp#L186-L192

my guess is that the 0x00 is missing at the end of the string and as a result gabage from the RAM until the next 0x00 is send too, but its hard to say without seeing the code.

**Comment by VaAndCob** [comment]
> do you have UTF-8 in you string / char array that you send via `broadcastTXT`? `broadcastTXT` does not alter you text / data by adding random strings. are you sure you data is 0x00 terminated?
> 
> the length is detected via `strlen` if you use use `broadcastTXT` with out the length parameter.
> 
> https://github.com/Links2004/arduinoWebSockets/blob/751cf87b6cd684c9d339f0314a18b0ee866d449c/src/WebSocketsServer.cpp#L186-L192
> 
> my guess is that the 0x00 is missing at the end of the string and as a result gabage from the RAM until the next 0x00 is send too, but its hard to say without seeing the code.

here is the code:
String content ="";
 content.concat("\nSong  | ");
  content.concat(song_title);
  content.concat("\nArtist| ");
  webSocket.broadcastTXT(content);
  
  just simple like this.  I notice there are terminate 0x00 characters at the end of  variable Artist.  this trigger decode text frame decoding error so often.  but when I add sanitize code by remove terminate 0x00,  it's now error rarely occur but still happen for sometimes.
  
  here is sanitize code to remove terminate 0x00
  
String sanitizeToUTF8(const String &input) {
  String sanitized = "";
  for (size_t i = 0; i < input.length(); i++) {
    //uint8_t c = input[i];
    char c = input.charAt(i);
    if (c != '\0') { //remove NULL
      sanitized += c;
    }
  }
  return sanitized;
}

song_title = sanitizeToUTF8(song_title);

How can I simply send string (around 360 bytes in length) and handle UTF-8 encoding?  Do I have to encode it myself or the library automatically encode sending string?

Thank you

**Comment by VaAndCob** [comment]
I dig into Websockets.cpp   the library handles UTF-8 encoding  , textframe but not sure if I have to include null byte 0x00 to the payload (string being send) manually? 
- as I read document. 
"If you're working with Arduino's String class, it automatically manages null-terminated strings internally, so you don't need to manually add a null terminator. "



content = content +

**Comment by Links2004** [comment]
the lib / broadcastTXT does not alter any Strings send.
as a result any invalide UTF-8 or control codes that you put in the function will be send as is.
in your case sanitizeing your input is the way to go.

yes `String` will be adding the needed null termination automatically at the end,
any other type that you can use with `broadcastTXT` will need a 0x00 at the end.

to simply remove any non Printable chars you can try
```cpp
String sanitizeToUTF8(const String &input) {
  String sanitized = "";
  for (size_t i = 0; i < input.length(); i++) {
  char c = input.charAt(i);
  if (isPrintable(c)) { // remove any non Printable chars
    sanitized += c;
  }
}
return sanitized;
}
```
 
for other filters take a look at https://www.arduino.cc/reference/en/ "Characters"
or
https://github.com/esp8266/Arduino/blob/f2da54d3a23bff4ca481d058b61c2b90a77ab3f1/cores/esp8266/WCharacter.h#L91-L93

**Comment by VaAndCob** [comment]
it works very well, no more errors,  but....   it also eliminates all Asian letters.   because I wanna send Artist and Track TItle over WebSocket,  but only English letters can be sent,  
Could u please advise me?  Thank you

**Comment by Links2004** [comment]
its not as easy since arduino is not realy desigend for UTF-8.
do you control the client side too?
if yes send the data as base64 and deal with UTF-8 on the client side (the browser knows better what UTF-8 is ;)

```cpp
bsong_title = base64.encode(song_title);
```

**Comment by VaAndCob** [comment]
Let me update:
now everything works like a charm, nice font for every language, correct indent ,

I have tried base64 data but have had no luck,  the library has duplicated references. cannot compile. So I try another way by 

I encode the string to UTF-8 with this function. the output is 3 HEX byte that represents 1 character. and broadcast as text.
`String encodeToUTF8(const String &input) {
  String sanitized = "";
  for (size_t i = 0; i < input.length()-1; i++) {//remove suffix 00 by minus length with 1
    char c = input.charAt(i);
    if (isPrintable(c) && c != '\0') { // Remove NULL and printable characters
      sanitized += c;//skip ascii char
    } else {
      char hexString[3];   // 2 characters for the hex representation + 1 for the null terminator
      snprintf(hexString, sizeof(hexString), "%02X", (unsigned char)c); // Convert character to hex and store in hexString
      sanitized += hexString;
    }
  }
  return sanitized;
}`

the client side I add UTF-8 decoding back to characters. and display. a bit slow but everything works well.  
`//Function decodeUTF-8 to character E0E881 -> "ก"
function decodeAndReplaceUTF8(inputString) {
  const regex = /([0-9A-Fa-f]{6})/g;
  const decodedString = inputString.replace(regex, (match, hexBytes) => {
    const byte1 = parseInt(hexBytes.substr(0, 2), 16);
    const byte2 = parseInt(hexBytes.substr(2, 2), 16);
    const byte3 = parseInt(hexBytes.substr(4, 2), 16);
    const utf8Character = String.fromCharCode(((byte1 & 0x0F) << 12) | ((byte2 & 0x3F) << 6) | (byte3 & 0x3F));
    return utf8Character;
  });
  return decodedString;
}`

In conclusion, it's like all characters were sent as printable (readable) ASCII characters.  so there is no trouble with decoding anymore.

Thank you for point me out the way.
